arangod/Indexes/IndexFactory.cpp

```diff
@@ -25,11 +25,129 @@
 #include "Basics/Exceptions.h"
 #include "Basics/StringUtils.h"
 #include "Basics/VelocyPackHelper.h"
+#include "Cluster/ServerState.h"
 #include "Indexes/Index.h"
 
 #include <velocypack/Slice.h>
 
-using namespace arangodb;
+namespace arangodb {
+
+Result IndexFactory::emplaceFactory(
+  std::string const& type,
+  IndexTypeFactory const& factory
+) {
+  if (!factory) {
+    return arangodb::Result(
+      TRI_ERROR_BAD_PARAMETER,
+      std::string("index factory undefined during index factory registration for index type '") + type + "'"
+    );
+  }
+
+  if (!_factories.emplace(type, factory).second) {
+    return arangodb::Result(
+      TRI_ERROR_ARANGO_DUPLICATE_IDENTIFIER,
+      std::string("index factory previously registered during index factory registration for index type '") + type + "'"
+    );
+  }
+
+  return arangodb::Result();
+}
+
+Result IndexFactory::emplaceNormalizer(
+  std::string const& type,
+  IndexNormalizer const& normalizer
+) {
+  if (!normalizer) {
```

Comment:
why not just use TRI_ASSERT ?

Reply:
because a TRI_ASSERT fails the entire app immediately, whereas there might be a valid usecase where the user might wish to handle this situation or ignore it altogether